File: backend/app/schemas/memory.py
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field, field_validator


MemoryCategory = Literal[
    "asset",
    "debt",
    "expense",
    "goal",
    "income",
    "preference",
    "profile",
    "other",
]
MemorySource = Literal["chat", "manual"]


class MemoryRequest(BaseModel):
    fact: str = Field(min_length=1, max_length=1000)
    category: MemoryCategory = "other"

    @field_validator("fact")
    @classmethod
    def normalize_fact(cls, value: str) -> str:
        value = " ".join(value.strip().split())
        if not value:
            raise ValueError("Memory fact must not be blank.")
        return value


class MemoryResponse(MemoryRequest):
    model_config = ConfigDict(from_attributes=True)

    id: int
    source: MemorySource
    created_at: datetime
    updated_at: datetime
    last_used_at: datetime | None = None


class MemoryExportResponse(BaseModel):
    generated_at: datetime
    memories: list[MemoryResponse]

File: frontend/src/api/memory.ts
import { apiRequest } from "./client";

export type MemoryCategory =
  | "asset"
  | "debt"
  | "expense"
  | "goal"
  | "income"
  | "preference"
  | "profile"
  | "other";

export type Memory = {
  id: number;
  fact: string;
  category: MemoryCategory;
  source: "chat" | "manual";
  created_at: string;
  updated_at: string;
  last_used_at?: string | null;
};

export type MemoryPayload = {
  fact: string;
  category: MemoryCategory;
};

export type MemoryExport = {
  generated_at: string;
  memories: Memory[];
};

export const getMemories = () =>
  apiRequest<Memory[]>("/memory", { authenticated: true });

export const exportMemories = () =>
  apiRequest<MemoryExport>("/memory/export", { authenticated: true });

export const createMemory = (payload: MemoryPayload) =>
  apiRequest<Memory>("/memory", {
    method: "POST",
    authenticated: true,
    body: JSON.stringify(payload),
  });

export const updateMemory = (id: number, payload: MemoryPayload) =>
  apiRequest<Memory>(`/memory/${id}`, {
    method: "PUT",
    authenticated: true,
    body: JSON.stringify(payload),
  });

export const deleteMemory = (id: number) =>
  apiRequest<void>(`/memory/${id}`, {
    method: "DELETE",
    authenticated: true,
  });
